File: gateways/python/stackverse_gateway/security.py
from __future__ import annotations

import hmac
import secrets
from urllib.parse import urlsplit

from starlette.requests import Request
from starlette.responses import Response

XSRF_COOKIE = "XSRF-TOKEN"
XSRF_HEADER = "x-xsrf-token"
SESSION_COOKIE = "stackverse_session"
CONTENT_SECURITY_POLICY = "default-src 'self'; base-uri 'self'; object-src 'none'; frame-ancestors 'none'"
STRICT_TRANSPORT_SECURITY = "max-age=31536000; includeSubDomains"

_STATE_CHANGING = {"POST", "PUT", "PATCH", "DELETE"}


def random_token(bytes_count: int = 32) -> str:
    return secrets.token_urlsafe(bytes_count)


def issue_csrf_cookie(request: Request, response: Response, secure: bool) -> None:
    if request.cookies.get(XSRF_COOKIE):
        return
    response.set_cookie(
        XSRF_COOKIE,
        random_token(16),
        httponly=False,
        secure=secure,
        samesite="lax",
        path="/",
    )


def has_valid_csrf(request: Request) -> bool:
    if request.method.upper() not in _STATE_CHANGING:
        return True
    cookie = request.cookies.get(XSRF_COOKIE)
    header = request.headers.get(XSRF_HEADER)
    return bool(cookie and header and hmac.compare_digest(cookie, header))


def is_same_origin_state_change(request: Request, expected_origin: str) -> bool:
    if not request.url.path.startswith("/api") or request.method.upper() not in _STATE_CHANGING:
        return True

    origin = request.headers.get("origin")
    if origin and _canonical_origin_or_none(origin) != expected_origin:
        return False

    fetch_site = request.headers.get("sec-fetch-site")
    return not fetch_site or fetch_site.lower() in {"same-origin", "none"}


def canonical_public_origin(public_url: str) -> str:
    return _canonical_origin(public_url)


def apply_security_headers(request: Request, response: Response, https_public_mode: bool) -> None:
    api_response = request.url.path.startswith("/api")
    response.headers["X-Content-Type-Options"] = "nosniff"
    if https_public_mode:
        response.headers["Strict-Transport-Security"] = STRICT_TRANSPORT_SECURITY
    if api_response:
        return

    response.headers["Referrer-Policy"] = "same-origin"
    response.headers["Content-Security-Policy"] = CONTENT_SECURITY_POLICY
    response.headers["X-Frame-Options"] = "DENY"
    response.headers["Cross-Origin-Opener-Policy"] = "same-origin"
    response.headers["Cross-Origin-Resource-Policy"] = "same-origin"


def _canonical_origin_or_none(value: str) -> str | None:
    if value.endswith("/"):
        return None
    try:
        origin = _canonical_origin(value)
    except ValueError:
        return None
    return origin if value == origin else None


def _canonical_origin(value: str) -> str:
    parts = urlsplit(value)
    if not parts.scheme or not parts.hostname or parts.path not in {"", "/"} or parts.query or parts.fragment:
        raise ValueError("origin must be a bare scheme/host/port")

    scheme = parts.scheme.lower()
    host = parts.hostname.lower()
    if ":" in host and not host.startswith("["):
        host = f"[{host}]"
    default_port = (scheme == "http" and parts.port in {None, 80}) or (scheme == "https" and parts.port in {None, 443})
    return f"{scheme}://{host}{'' if default_port else f':{parts.port}'}"

File: gateways/node-fastify/src/security.ts
import { randomBytes, timingSafeEqual } from "node:crypto";
import type { FastifyReply, FastifyRequest } from "fastify";

export const XSRF_COOKIE = "XSRF-TOKEN";
export const XSRF_HEADER = "x-xsrf-token";
export const SESSION_COOKIE = "stackverse_session";

export const CONTENT_SECURITY_POLICY = "default-src 'self'; base-uri 'self'; object-src 'none'; frame-ancestors 'none'";
export const STRICT_TRANSPORT_SECURITY = "max-age=31536000; includeSubDomains";

const STATE_CHANGING = new Set(["POST", "PUT", "PATCH", "DELETE"]);

export function randomToken(bytes = 32): string {
  return randomBytes(bytes).toString("base64url");
}

export function issueCsrfCookie(request: FastifyRequest, reply: FastifyReply, secure: boolean): void {
  if (request.cookies[XSRF_COOKIE]) return;
  reply.setCookie(XSRF_COOKIE, randomToken(16), {
    httpOnly: false,
    sameSite: "lax",
    secure,
    path: "/",
  });
}

export function hasValidCsrf(request: FastifyRequest): boolean {
  if (!STATE_CHANGING.has(request.method.toUpperCase())) return true;

  const cookie = request.cookies[XSRF_COOKIE];
  const header = headerValue(request, XSRF_HEADER);
  if (!cookie || !header) return false;

  const cookieBytes = Buffer.from(cookie, "utf8");
  const headerBytes = Buffer.from(header, "utf8");
  return cookieBytes.length === headerBytes.length && timingSafeEqual(cookieBytes, headerBytes);
}

export function isSameOriginStateChange(request: FastifyRequest, expectedOrigin: string): boolean {
  if (!request.url.startsWith("/api") || !STATE_CHANGING.has(request.method.toUpperCase())) {
    return true;
  }

  const origin = headerValue(request, "origin");
  if (origin && canonicalOriginOrNull(origin) !== expectedOrigin) {
    return false;
  }

  const fetchSite = headerValue(request, "sec-fetch-site");
  return !fetchSite || fetchSite.toLowerCase() === "same-origin" || fetchSite.toLowerCase() === "none";
}

export function canonicalPublicOrigin(publicUrl: URL): string {
  return canonicalOrigin(publicUrl);
}

export function applySecurityHeaders(request: FastifyRequest, reply: FastifyReply, httpsPublicMode: boolean): void {
  const apiResponse = request.url.startsWith("/api");
  reply.header("X-Content-Type-Options", "nosniff");
  if (httpsPublicMode) {
    reply.header("Strict-Transport-Security", STRICT_TRANSPORT_SECURITY);
  }

  if (apiResponse) return;

  reply.header("Referrer-Policy", "same-origin");
  reply.header("Content-Security-Policy", CONTENT_SECURITY_POLICY);
  reply.header("X-Frame-Options", "DENY");
  reply.header("Cross-Origin-Opener-Policy", "same-origin");
  reply.header("Cross-Origin-Resource-Policy", "same-origin");
}

export function headerValue(request: FastifyRequest, name: string): string | undefined {
  const value = request.headers[name.toLowerCase()];
  if (Array.isArray(value)) return value[0];
  return value;
}

function canonicalOriginOrNull(value: string): string | null {
  let url: URL;
  try {
    url = new URL(value);
  } catch {
    return null;
  }
  if (url.search || url.hash || value.endsWith("/")) {
    return null;
  }
  const origin = canonicalOrigin(url);
  return value === origin ? origin : null;
}

function canonicalOrigin(url: URL): string {
  if (!url.protocol || !url.hostname) {
    throw new Error("PUBLIC_URL must include a scheme and host.");
  }
  const scheme = url.protocol.slice(0, -1).toLowerCase();
  let host = url.hostname.toLowerCase();
  if (host.includes(":") && !host.startsWith("[")) {
    host = `[${host}]`;
  }
  const defaultPort =
    (scheme === "http" && (url.port === "" || url.port === "80")) ||
    (scheme === "https" && (url.port === "" || url.port === "443"));
  return `${scheme}://${host}${defaultPort ? "" : `:${url.port}`}`;
}
